pdi-execution-engine/api/src/main/java/org/pentaho/di/engine/api/IExecutionContext.java

```diff
@@ -1,17 +1,26 @@
 package org.pentaho.di.engine.api;
 
+import org.pentaho.di.engine.api.reporting.IProgressReporting;
+
 import java.util.Map;
+import java.util.concurrent.CompletableFuture;
 
 /**
  * Created by nbaker on 5/31/16.
  */
-public interface IExecutionContext {
-
+public interface IExecutionContext extends IProgressReporting {
   Map<String, Object> getParameters();
 
   Map<String, Object> getEnvironment();
 
   ITransformation getTransformation();
 
+  /**
+   * Transformation parameters are the preferred way of controlling execution.
+   * Maybe arguments can be rolled into environment?
+   */
+  @Deprecated
   String[] getArguments();
+
+  CompletableFuture<IExecutionResult> execute();
```

Comment:
I prefer the Java 8 CompletableFuture, since it allows for callbacks in addition to blocking. Could switch to a CompletionStage if you prefer
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html